rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/ParametricTextDocumentService.java

```diff
@@ -372,6 +386,69 @@ private CodeLens locCommandTupleToCodeLense(String languageName, IValue v) {
 
         return new CodeLens(Locations.toRange(loc, columns), constructorToCommand(languageName, command), null);
     }
+    
+    private CodeAction constructorToCodeAction(String languageName, IConstructor codeAction) {
+        IWithKeywordParameters<?> kw = codeAction.asWithKeywordParameters();
+        IConstructor command = (IConstructor) kw.getParameter("command");
+        IString title = (IString) kw.getParameter("title");
+        IList edits = (IList) kw.getParameter("edits");
+        IConstructor kind = (IConstructor) kw.getParameter("kind");
+
+        // first deal with the defaults. Must mimick what's in util::LanguageServer with the `data CodeAction` declaration
+        if (title == null) {
+            if (command != null) {
+                title = (IString) command.asWithKeywordParameters().getParameter("title");
+            }
+
+            if (title == null) {
+                title = IRascalValueFactory.getInstance().string("");
+            }
+        }
+
+        CodeAction result = new CodeAction(title.getValue());
+
+        if (command != null) {
+            result.setCommand(constructorToCommand(languageName, command));
+        }
+
+        if (edits != null) {
+            result.setEdit(new WorkspaceEdit(DocumentChanges.translateDocumentChanges(this, edits)));
+        }
+        
+        result.setKind(constructorToCodeActionKind(kind));
+
+        return result;
+    }
+
+    /**
+     * Translates `refactor(inline())` to `"refactor.inline"` and `empty()` to `""`, etc.
+     */
+    private String constructorToCodeActionKind(IConstructor kind) {
+        if (kind == null) {
```

Comment:
can it be null? if so, add `@Nullable` to the parameter.

Reply:
Yes this is Nullable.